// home/actions.js
export const CONNECT_TO_HOME = "CONNECT_TO_HOME";
export const connectToHome = (location, login, password) => ({
    type: CONNECT_TO_HOME,
    payload: {location, login, password}
});

export const CONNECTION_TO_HOME_FAILED = "CONNECTION_TO_HOME_FAILED";
export const connectionToHomeFailed = () => ({
    type: CONNECTION_TO_HOME_FAILED
});

export const CONNECTED_TO_HOME = "CONNECTED_TO_HOME";
export const connectedToHome = (location, login, token) => ({
    type: CONNECTED_TO_HOME,
    payload: {location, login, token}
});

export const DISCONNECT_FROM_HOME = "DISCONNECT_FROM_HOME";
export const disconnectFromHome = (location) => ({
    type: DISCONNECT_FROM_HOME,
    payload: {location}
});

export const HOME_OWNER_SET = "HOME_OWNER_SET";
export const homeOwnerSet = (name, generation) => ({
    type: HOME_OWNER_SET,
    payload: {name, generation}
});

export const HOME_OWNER_VERIFIED = "HOME_OWNER_VERIFIED";
export const homeOwnerVerified = (name, generation, latest, correct) => ({
    type: HOME_OWNER_VERIFIED,
    payload: {name, generation, latest, correct}
});
