// home/connect.js
import { call, put, select } from 'redux-saga/effects';

import { messageBox } from "messagebox/actions";
import { normalizeUrl } from "util/util";
import { connectedToHome, connectionToHomeFailed, homeOwnerSet, homeOwnerVerified } from "home/actions";
import { openConnectDialog } from "logobar/connectionstatus/connectdialog/actions";
import * as NodeApi from "node/api";
import { callApi } from "node/sagas";
import * as Home from "home/sagas";
import * as Naming from "naming/sagas";
import { NodeApiError } from "error";
import { errorThrown } from "error/actions";

function* connectToHomeFailure(error, onClose = null) {
    yield put(connectionToHomeFailed());
    yield put(messageBox("Connection to home failed: " + error.message, onClose));
}

export function* connectToHomeSaga(action) {
    const {location, login, password} = action.payload;
    let data;
    try {
        data = yield call(callApi, {
            location: "/tokens",
            rootApiSelector: normalizeUrl(location) + "/moera/api",
            method: "POST",
            body: {
                login,
                password
            },
            schema: NodeApi.TokenCreated,
            errorFilter: ["credentials.login-incorrect"]
        });
    } catch (e) {
        if (e instanceof NodeApiError) {
            yield call(connectToHomeFailure, e, openConnectDialog());
        } else {
            yield call(connectToHomeFailure, e);
        }
        return;
    }
    yield put(connectedToHome(normalizeUrl(location), login, data.token, data.permissions));
    yield call(verifyHomeOwnerSaga);
}

export function* verifyHomeOwnerSaga() {
    try {
        const data = yield call(Home.getWhoAmI);
        yield put(homeOwnerSet(data.registeredName, data.registeredNameGeneration));
        if (data.registeredName == null) {
            return;
        }
        const ndata = yield call(Naming.getCurrent, data.registeredName, data.registeredNameGeneration);
        const rootLocation = yield select(state => state.node.root.location);
        const correct = ndata && normalizeUrl(ndata.nodeUri) === rootLocation;
        const latest = ndata && ndata.latest;
        yield put(homeOwnerVerified(data.registeredName, data.registeredNameGeneration, latest, correct));
    } catch (e) {
        yield put(errorThrown(e));
    }
}
