// home/reducer.js
import {
    CONNECT_TO_HOME,
    CONNECTED_TO_HOME,
    CONNECTION_TO_HOME_FAILED, DISCONNECT_FROM_HOME
} from "home/actions";

const initialState = {
    connecting: false,
    root: {
        location: null,
        page: null,
        api: null
    },
    token: null,
    login: null
};

export default (state = initialState, action) => {
    switch (action.type) {
        case CONNECT_TO_HOME:
            return {
                ...state,
                connecting: true
            };

        case CONNECTION_TO_HOME_FAILED:
            return {
                ...state,
                connecting: false
            };

        case CONNECTED_TO_HOME:
            return {
                ...state,
                connecting: false,
                root: {
                    location: action.payload.location,
                    page: action.payload.location + "/moera",
                    api: action.payload.location + "/moera/api",
                },
                login: action.payload.login,
                token: action.payload.token
            };

        case DISCONNECT_FROM_HOME:
            return {
                ...state,
                root: {
                    location: null,
                    page: null,
                    api: null
                },
                token: null,
                login: null
            };

        default:
            return state;
    }
}
