// state/timeline/actions.js
export const TIMELINE_PAST_SLICE_LOAD = "TIMELINE_PAST_SLICE_LOAD";
export const timelinePastSliceLoad = () => ({
    type: TIMELINE_PAST_SLICE_LOAD
});

export const TIMELINE_PAST_SLICE_LOAD_FAILED = "TIMELINE_PAST_SLICE_LOAD_FAILED";
export const timelinePastSliceLoadFailed = () => ({
    type: TIMELINE_PAST_SLICE_LOAD_FAILED
});

export const TIMELINE_FUTURE_SLICE_LOAD = "TIMELINE_FUTURE_SLICE_LOAD";
export const timelineFutureSliceLoad = () => ({
    type: TIMELINE_FUTURE_SLICE_LOAD
});

export const TIMELINE_FUTURE_SLICE_LOAD_FAILED = "TIMELINE_FUTURE_SLICE_LOAD_FAILED";
export const timelineFutureSliceLoadFailed = () => ({
    type: TIMELINE_FUTURE_SLICE_LOAD_FAILED
});

export const TIMELINE_PAST_SLICE_SET = "TIMELINE_PAST_SLICE_SET";
export const timelinePastSliceSet = (postings, before, after) => ({
    type: TIMELINE_PAST_SLICE_SET,
    payload: {postings, before, after}
});

export const TIMELINE_FUTURE_SLICE_SET = "TIMELINE_FUTURE_SLICE_SET";
export const timelineFutureSliceSet = (postings, before, after) => ({
    type: TIMELINE_FUTURE_SLICE_SET,
    payload: {postings, before, after}
});

export const TIMELINE_SCROLLED = "TIMELINE_SCROLLED";
export const timelineScrolled = (at) => ({
    type: TIMELINE_SCROLLED,
    payload: {at}
});
