// state/trigger.js
import { put, select, takeEvery } from 'redux-saga/effects';

export function trigger(signal, filter, action) {
    return {signal, filter, action};
}

function addTrigger(triggers, trigger) {
    const {signal, filter, action} = trigger;

    if (filter === false) {
        return;
    }
    if (Array.isArray(signal)) {
        for (let sg of signal) {
            addTrigger(triggers, {signal: sg, filter, action});
        }
        return;
    }
    let list = triggers.get(signal);
    if (list === undefined) {
        list = [];
    }
    list.push({filter, action});
    triggers.set(signal, list);
}

export function collectTriggers(...lists) {
    let triggers = new Map();
    for (let list of lists) {
        if (Array.isArray(list)) {
            for (let trigger of list) {
                addTrigger(triggers, trigger);
            }
        } else {
            addTrigger(triggers, list);
        }
    }
    return triggers;
}

function* triggersSaga(triggers, {type: signal}) {
    let list = triggers.get(signal);
    if (list === undefined) {
        return;
    }
    for (let trigger of list) {
        let enabled;
        if (typeof(trigger.filter) === "function") {
            enabled = !!(yield select(trigger.filter, signal));
        } else {
            enabled = !!trigger.filter;
        }
        if (enabled) {
            if (typeof(trigger.action) === "function") {
                yield put(trigger.action(signal));
            } else {
                yield put(trigger.action);
            }
        }
    }
}

export function* invokeTriggers(triggers) {
    yield takeEvery([...triggers.keys()], triggersSaga, triggers);
}

export function inv(func) {
    return function (...params) {
        return !func(...params);
    }
}

export function conj(...funcs) {
    if (!Array.isArray(funcs)) {
        return funcs;
    }
    return function (...params) {
        for (let func of funcs) {
            if (!func(...params)) {
                return false;
            }
        }
        return true;
    }
}

export function disj(...funcs) {
    if (!Array.isArray(funcs)) {
        return funcs;
    }
    return function (...params) {
        for (let func of funcs) {
            if (func(...params)) {
                return true;
            }
        }
        return false;
    }
}
