// ui/navigation/LocationInfo.js
export default class LocationInfo {

    constructor() {
        this.directories = [];
        this.parameters = {};
        this.title = "";
    }

    clone() {
        let info = new LocationInfo();
        info.directories = [...this.directories];
        info.parameters = {...this.parameters};
        info.title = this.title;
        return info;
    }

    sub(name) {
        let info = this.clone();
        info.directories.push(name);
        return info;
    }

    withParameter(name, value) {
        let info = this.clone();
        info.parameters[name] = value;
        return info;
    }

    withTitle(title) {
        let info = this.clone();
        info.title = title;
        return info;
    }

    hasTitle() {
        return this.title != null && this.title !== "";
    }

    toUrl() {
        let path = this.directories.map(dir => "/" + dir).join("");
        let query = "";
        for (let name in this.parameters) {
            if (this.parameters.hasOwnProperty(name)) {
                query += (query === "" ? "?" : "&") + name + "=" + encodeURIComponent(this.parameters[name]);
            }
        }
        return path + query;
    }

}
