// ui/navigation/Navigation.js
import React from 'react';
import { connect } from 'react-redux';

import { goToLocation } from "state/navigation/actions";

class Navigation extends React.Component {

    constructor(props) {
        super(props);

        this.popState = this.popState.bind(this);
    }

    componentDidMount() {
        window.onpopstate = this.popState;
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        const {rootPage, location, title} = this.props;

        if (location !== prevProps.location) {
            window.history.pushState({location}, "", rootPage + location);
        }
        if (title !== prevProps.title) {
            if (title) {
                document.title = title + " | Moera";
            } else {
                document.title = "Moera";
            }
        }
    }

    popState(event) {
        this.props.goToLocation(window.location.pathname, window.location.search);
        event.preventDefault();
    }

    render() {
        return null;
    }

}

export default connect(
    (state) => ({
        rootPage: state.node.root.page,
        location: state.navigation.location,
        title: state.navigation.title
    }),
    { goToLocation }
)(Navigation);
