// profile/Profile.js
import React from 'react';
import PropType from 'prop-types';
import { connect } from 'react-redux';

import { Page } from "page/Page";
import { Button, Loading, RegisteredName } from "control";
import { profileLoad } from "profile/actions";
import { isPermitted } from "util/permissions";
import { getHomePermissions } from "home/selectors";

const EditButton = () => (
    <Button variant="outline-primary" size="sm" style={{"margin-left": "1rem"}}>Edit</Button>
);

const ProfileLine = ({title, children, visible = true}) => (
    visible &&
        <>
            <dt className="col-sm-2">{title}</dt>
            <dd className="col-sm-10">{children}</dd>
        </>
);

ProfileLine.propTypes = {
    title: PropType.string,
    visible: PropType.bool
};

class Profile extends React.PureComponent {

    static propTypes = {
        loading: PropType.bool,
        registeredName: PropType.string,
        registeredNameGeneration: PropType.number,
        fullName: PropType.string,
        gender: PropType.string
    };

    componentDidMount() {
        this.props.profileLoad();
    }

    render() {
        const {
            loading,
            registeredName,
            registeredNameGeneration,
            fullName,
            gender,
            editable} = this.props;

        return (
            <Page>
                <h2>
                    Profile <Loading active={loading} />
                    {editable && <EditButton/>}
                </h2>
                <dl className="row">
                    <ProfileLine title="Registered name">
                        <RegisteredName name={registeredName} generation={registeredNameGeneration}/>
                    </ProfileLine>
                    <ProfileLine title="Full name" visible={!!fullName}>
                        {fullName}
                    </ProfileLine>
                    <ProfileLine title="Gender" visible={!!gender}>
                        {gender}
                    </ProfileLine>
                </dl>
            </Page>
        );
    }

}

export default connect(
    (state) => ({
        ...state.profile,
        editable: !state.profile.loading
            && isPermitted("edit", state.profile.operations, getHomePermissions(state), state)
    }),
    { profileLoad }
)(Profile);
