// profile/reducer.js
import { PROFILE_EDIT, PROFILE_EDIT_CANCEL, PROFILE_LOAD, PROFILE_LOAD_FAILED, PROFILE_SET } from "profile/actions";

const initialState = {
    loading: false,
    registeredName: null,
    registeredNameGeneration: null,
    fullName: null,
    gender: null,
    operations: {
        edit: []
    },
    editing: false,
    profileEditor: {
        fullName: null,
        gender: null
    }
};

export default (state = initialState, action) => {
    switch (action.type) {
        case PROFILE_LOAD:
            return {
                ...state,
                loading: true
            };

        case PROFILE_SET:
            return {
                ...state,
                ...action.payload.profile,
                loading: false
            };

        case PROFILE_LOAD_FAILED:
            return {
                ...state,
                loading: false
            };

        case PROFILE_EDIT:
            return {
                ...state,
                editing: true,
                profileEditor: {
                    ...state.profileEditor,
                    fullName: state.fullName,
                    gender: state.gender
                }
            };

        case PROFILE_EDIT_CANCEL:
            return {
                ...state,
                editing: false
            };

        default:
            return state;
    }
}
