// api/naming/api.js
import schema from "api/schema";

export const Result = schema({
    type: "object",
    properties: {
        "jsonrpc": {
            type: "string"
        },
        "result": {
            type: "object"
        },
        "id": {
            type: "integer"
        }
    },
    additionalProperties: false,
    required: ["jsonrpc", "result", "id"]
});

export const ErrorResult = schema({
    type: "object",
    properties: {
        "jsonrpc": {
            type: "string"
        },
        "error": {
            type: "object",
            "code": {
                type: "integer"
            },
            "message": {
                type: "string"
            },
            additionalProperties: false,
            required: ["code", "message"]
        },
        "id": {
            type: "integer"
        }
    },
    additionalProperties: false,
    required: ["jsonrpc", "error", "id"]
});

export const RegisteredNameInfo = schema({
    type: "object",
    properties: {
        "name": {
            type: "string"
        },
        "generation": {
            type: "integer",
            minimum: 0
        },
        "latest": {
            type: "boolean"
        },
        "updatingKey": {
            type: "string"
        },
        "nodeUri": {
            type: "string"
        },
        "deadline": {
            type: "integer"
        },
        "signingKey": {
            type: "string"
        },
        "validFrom": {
            type: "integer"
        }
    },
    additionalProperties: false,
    required: ["name", "generation", "latest", "updatingKey", "deadline"]
});
