// node/api.js
import schema from "util/schema";

export const Result = schema({
    type: "object",
    properties: {
        "errorCode": {
            type: "string"
        },
        "message": {
            type: "string",
            default: ""
        }
    },
    additionalProperties: false,
    required: ["errorCode", "message"]
});

export const WhoAmI = schema({
    type: "object",
    properties: {
        "registeredName": {
            type: "string"
        },
        "registeredNameGeneration": {
            type: "integer"
        }
    },
    additionalProperties: false,
});

export const ProfileInfo = schema({
    type: "object",
    properties: {
        "registeredName": {
            type: "string"
        },
        "registeredNameGeneration": {
            type: "integer"
        },
        "fullName": {
            type: "string"
        },
        "gender": {
            type: "string"
        },
        "operations": {
            type: "object",
            properties: {
                "edit": {
                    type: "array",
                    items: {
                        type: "string"
                    }
                }
            },
            additionalProperties: false,
        }
    },
    additionalProperties: false,
});

export const TokenCreated = schema({
    type: "object",
    properties: {
        "token": {
            type: "string"
        },
        "permissions": {
            type: "array",
            items: {
                type: "string"
            },
            default: []
        }
    },
    additionalProperties: false,
    required: ["token", "permissions"]
});
