// ui/events/Events.js
import React from 'react';
import { Client } from '@stomp/stompjs';
import { Event, formatSchemaErrors } from "api";

export default class Events extends React.Component {

    constructor(props) {
        super(props);

        this.stomp = null;
        this.location = null;
        this.queueStartedAt = null;
        this.lastEvent = null;
        this.buffer = [];

        this.onConnect = this.onConnect.bind(this);
        this.onMessage = this.onMessage.bind(this);
    }

    componentDidMount() {
        this._connect();
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (prevProps.location !== this.props.location) {
            this._disconnect();
            this._connect();
        }
    }

    _connect() {
        const {location} = this.props;

        if (location == null) {
            return;
        }
        if (location !== this.location) {
            this.location = location;
            this.queueStartedAt = null;
            this.lastEvent = null;
        }

        this.stomp = new Client({
            brokerURL: location,
            onConnect: this.onConnect,
        });
        this.stomp.activate();
    }

    _disconnect() {
        if (this.stomp != null) {
            this.stomp.deactivate();
            this.stomp = null;
        }
    }

    onConnect() {
        let headers = {};
        if (this.queueStartedAt != null) {
            headers.seen = `${this.queueStartedAt},${this.lastEvent}`;
        }
        this.queueStartedAt = 0;
        this.lastEvent = 0;
        this.buffer = [];
        this.stomp.subscribe("/user/queue", this.onMessage, headers);
    }

    onMessage(message) {
        let event = JSON.parse(message.body);
        if (Event(message.body)) {
            console.error("Incorrect event received", formatSchemaErrors(event.errors));
        }
        this.buffer.push(event);
        this.buffer.sort((e1, e2) => e2.ordinal - e1.ordinal);
        while (this.buffer.length > 0
                && (this.lastEvent === 0 || this.buffer.length > 5 || this.buffer[0].ordinal <= this.lastEvent + 1)) {
            console.log("Event:", this.buffer[0]);
            this.queueStartedAt = this.buffer[0].queueStartedAt;
            this.lastEvent = this.buffer[0].ordinal;
            this.buffer.shift();
        }
    }

    render() {
        return null;
    }

}
