// util/duration.js
export class DurationParsingError extends Error {

    constructor(value) {
        super("Error parsing duration value: " + value);
    }

}

export default class Duration {

    constructor(amount, unit) {
        this.amount = amount;
        this.unit = unit;
    }

    static parse(v) {
        if (v == null || v.length === 0) {
            return new Duration(0, "s");
        }
        if (!/^[0-9]+[smhd]$/i.test(v)) {
            throw new DurationParsingError(v);
        }
        return new Duration(
            parseInt(v.substring(0, v.length - 1)),
            v.charAt(v.length - 1).toLowerCase()
        )
    }

    toSeconds() {
        let factor;
        switch (this.unit) {
            case "m":
                factor = 60;
                break;
            case "h":
                factor = 60 * 60;
                break;
            case "d":
                factor = 24 * 60 * 60;
                break;
            default:
                factor = 1;
        }
        return this.amount * factor;
    }

    toString() {
        return `${this.amount}${this.unit}`;
    }

}
