// util/util.js
import * as Base64js from 'base64-js';
import * as URI from 'uri-js';

export function normalizeUrl(url) {
    return url.endsWith("/") ? url.substring(0, url.length - 1) : url;
}

export function urlWithParameters(url, parameters) {
    let query = "";
    for (let name in parameters) {
        if (parameters.hasOwnProperty(name) && parameters[name] != null) {
            query += (query === "" ? "" : "&") + name + "=" + encodeURIComponent(parameters[name]);
        }
    }
    if (query === "") {
        return url;
    }
    return url + (url.indexOf("?") < 0 ? "?" : "&") + query;
}

export function toWsUrl(url) {
    let components = URI.parse(url);
    components.scheme = components.scheme.toLowerCase() === "https" ? "wss" : "ws";
    return URI.serialize(components);
}

export function atOwner(state) {
    return state.owner.name ? " @ " + state.owner.name : "";
}

export function range(length) {
    return [...Array(length).keys()];
}

export function randomId() {
    let buf = new Uint8Array(16);
    window.crypto.getRandomValues(buf);
    return Base64js.fromByteArray(buf);
}

export function parseBool(val) {
    const ival = val.toLowerCase();
    return ival === "yes" || ival === "true" || ival === "1";
}

export function now() {
    return Math.floor(Date.now() / 1000);
}
