// state/navigation/actions.js
import { PAGE_POSTING, PAGE_PROFILE, PAGE_TIMELINE } from "state/navigation/pages";

export const INIT_FROM_LOCATION = "INIT_FROM_LOCATION";
export const initFromLocation = (rootLocation, path, query) => ({
    type: INIT_FROM_LOCATION,
    payload: {rootLocation, path, query}
});

export const GO_TO_PAGE = "GO_TO_PAGE";
export const goToPage = (page, details = {}) => ({
    type: GO_TO_PAGE,
    payload: {page, details}
});
export const goToProfile = () => goToPage(PAGE_PROFILE);
export const goToTimeline = (at = null) => goToPage(PAGE_TIMELINE, {at});
export const goToPosting = (id) => goToPage(PAGE_POSTING, {id});

export const UPDATE_LOCATION = "UPDATE_LOCATION";
export const updateLocation = () => ({
    type: UPDATE_LOCATION
});

export const SCROLL_LOCATION = "SCROLL_LOCATION";
export const scrollLocation = () => ({
    type: SCROLL_LOCATION
});

export const LOCATION_SET = "LOCATION_SET";
export const locationSet = (location, title, update) => ({
    type: LOCATION_SET,
    payload: {location, title, update}
});

export const LOCATION_LOCK = "LOCATION_LOCK";
export const locationLock = () => ({
    type: LOCATION_LOCK
});

export const LOCATION_UNLOCK = "LOCATION_UNLOCK";
export const locationUnlock = () => ({
    type: LOCATION_UNLOCK
});

export const GO_TO_LOCATION = "GO_TO_LOCATION";
export const goToLocation = (path, query) => ({
    type: GO_TO_LOCATION,
    payload: {path, query}
});
