// ui/posting/PostingMenu.js
import React from 'react';
import { Manager, Popper, Reference } from 'react-popper';
import cx from 'classnames';

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import "./PostingMenu.css";

export default class PostingMenu extends React.PureComponent {

    constructor(props) {
        super(props);
        this.state = {visible: false};

        this.toggle = this.toggle.bind(this);
        this.hide = this.hide.bind(this);
    }

    toggle() {
        if (!this.state.visible) {
            this.show();
        } else {
            this.hide();
        }
    }

    show() {
        this.setState({visible: true});
        this.scheduleUpdate();
        document.addEventListener("click", this.hide);
    }

    hide() {
        this.setState({visible: false});
        this.scheduleUpdate();
        document.removeEventListener("click", this.hide);
    }

    render() {
        const pointerEvents = {pointerEvents: this.state.visible ? "auto" : "none"};
        return (
            <Manager>
                <Reference>
                    {({ref}) => (
                        <div className="menu" ref={ref} onClick={this.toggle}>
                            <FontAwesomeIcon icon="chevron-down" className="chevron" />
                            <Popper placement="bottom-end">
                                {({ref, style, placement, scheduleUpdate}) => {
                                    this.scheduleUpdate = scheduleUpdate;
                                    return (
                                        <div ref={ref} style={{...style, ...pointerEvents}} className={cx(
                                            `bs-popover-${placement}`,
                                            "fade",
                                            "dropdown-menu",
                                            {"show": this.state.visible}
                                        )}>
                                            <a className="dropdown-item" href="#">Edit...</a>
                                            <a className="dropdown-item" href="#">Delete</a>
                                        </div>
                                    );
                                }}
                            </Popper>
                        </div>
                        )}
                </Reference>
            </Manager>
        );
    }

}
