// storage/Storage.js
import React from 'react';
import { connect } from 'react-redux';
import deepEqual from 'deep-equal';

import { getHomeToken } from "home/selectors";
import { restoreConnectDialog } from "logobar/connectionstatus/connectdialog/actions";
import { restoreConnectionToHome } from "home/actions";

class Storage extends React.Component {

    constructor(props) {
        super(props);

        this.initialized = false;
        this.messageReceived = this.messageReceived.bind(this);
    }

    componentDidMount() {
        window.addEventListener("message", this.messageReceived);
    }

    messageReceived(event) {
        // Only accept messages from the same frame
        if (event.source !== window) {
            return;
        }

        const message = event.data;

        // Only accept messages that we know are ours
        if (message === null || typeof message !== "object") {
            return;
        }

        console.log("Message received by client: " + JSON.stringify(message));
        switch (message.action) {
            case "loadedData":
                this.loadedData(message.payload);
                this.initialized = true;
                return;

            default:
                return;
        }
    }

    loadedData(data) {
        if (!data) {
            return;
        }

        const {location, login, token} = data.home;
        this.props.restoreConnectDialog(location, login);
        if (token) {
            this.props.restoreConnectionToHome(location, login, token);
        }
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (!this.initialized || deepEqual(prevProps, this.props)) {
            return;
        }

        window.postMessage({
            action: "storeData",
            payload: {
                home: {...this.props.home}
            }
        }, "*");
    }

    render() {
        return null;
    }

}

export default connect(
    (state) => ({
        home: {
            location: state.home.root.location,
            login: state.home.login,
            token: getHomeToken(state)
        },
    }),
    { restoreConnectDialog, restoreConnectionToHome }
)(Storage);
