// control/Popover.js
import React from 'react';
import PropType from 'prop-types';
import cx from 'classnames';
import { Manager, Popper, Reference } from 'react-popper';

export class Popover extends React.PureComponent {

    constructor(props) {
        super(props);
        this.state = {visible: false};

        this.toggle = this.toggle.bind(this);
    }

    static propTypes = {
        text: PropType.string.required,
        textClassName: PropType.string
    };

    toggle() {
        this.setState({visible: !this.state.visible});
    }

    render() {
        return (
            <Manager>
                <Reference>
                    {({ref}) => (
                        <span ref={ref} onClick={this.toggle} className={this.props.textClassName}>
                            {this.props.text}
                        </span>
                    )}
                </Reference>
                <Popper placement="bottom">
                    {({ref, style, placement, arrowProps}) => (
                        <div ref={ref} style={style} className={cx([
                            "popover",
                            `bs-popover-${placement}`,
                            "fade",
                            {"show": this.state.visible}
                        ])}>
                            <div ref={arrowProps.ref} style={arrowProps.style} className="arrow"/>
                            <div className="popover-body">{this.props.children}</div>
                        </div>
                    )}
                </Popper>
            </Manager>
        );
    }
}
