// state/registeredname/actions.js
export const REGISTERED_NAME_LOAD = "REGISTERED_NAME_LOAD";
export const registeredNameLoad = () => ({
    type: REGISTERED_NAME_LOAD
});

export const REGISTERED_NAME_LOAD_FAILED = "REGISTERED_NAME_LOAD_FAILED";
export const registeredNameLoadFailed = () => ({
    type: REGISTERED_NAME_LOAD_FAILED
});

export const REGISTERED_NAME_SET = "REGISTERED_NAME_SET";
export const registeredNameSet = (registeredName) => ({
    type: REGISTERED_NAME_SET,
    payload: {registeredName}
});

export const REGISTERED_NAME_UNSET = "REGISTERED_NAME_UNSET";
export const registeredNameUnset = (registeredName) => ({
    type: REGISTERED_NAME_UNSET
});

export const REGISTER_NAME_DIALOG = "REGISTER_NAME_DIALOG";
export const registerNameDialog = () => ({
    type: REGISTER_NAME_DIALOG
});

export const REGISTER_NAME_DIALOG_CANCEL = "REGISTER_NAME_DIALOG_CANCEL";
export const registerNameDialogCancel = () => ({
    type: REGISTER_NAME_DIALOG_CANCEL
});

export const REGISTER_NAME = "REGISTER_NAME";
export const registerName = (name, formik) => ({
    type: REGISTER_NAME,
    payload: {name, formik}
});

export const REGISTER_NAME_SUCCEEDED = "REGISTER_NAME_SUCCEEDED";
export const registerNameSucceeded = () => ({
    type: REGISTER_NAME_SUCCEEDED
});

export const REGISTER_NAME_FAILED = "REGISTER_NAME_FAILED";
export const registerNameFailed = () => ({
    type: REGISTER_NAME_FAILED
});
