// state/registeredname/actions.js
export const REGISTERED_NAME_LOAD = "REGISTERED_NAME_LOAD";
export const registeredNameLoad = () => ({
    type: REGISTERED_NAME_LOAD
});

export const REGISTERED_NAME_LOAD_FAILED = "REGISTERED_NAME_LOAD_FAILED";
export const registeredNameLoadFailed = () => ({
    type: REGISTERED_NAME_LOAD_FAILED
});

export const REGISTERED_NAME_SET = "REGISTERED_NAME_SET";
export const registeredNameSet = (registeredName) => ({
    type: REGISTERED_NAME_SET,
    payload: {registeredName}
});

export const REGISTERED_NAME_UNSET = "REGISTERED_NAME_UNSET";
export const registeredNameUnset = (registeredName) => ({
    type: REGISTERED_NAME_UNSET
});

export const REGISTER_NAME_DIALOG = "REGISTER_NAME_DIALOG";
export const registerNameDialog = () => ({
    type: REGISTER_NAME_DIALOG
});

export const REGISTER_NAME_DIALOG_CANCEL = "REGISTER_NAME_DIALOG_CANCEL";
export const registerNameDialogCancel = () => ({
    type: REGISTER_NAME_DIALOG_CANCEL
});

export const REGISTER_NAME = "REGISTER_NAME";
export const registerName = (name, onNameTaken) => ({
    type: REGISTER_NAME,
    payload: {name, onNameTaken}
});

export const REGISTER_NAME_SUCCEEDED = "REGISTER_NAME_SUCCEEDED";
export const registerNameSucceeded = (name, mnemonic) => ({
    type: REGISTER_NAME_SUCCEEDED,
    payload: {name, mnemonic}
});

export const REGISTER_NAME_FAILED = "REGISTER_NAME_FAILED";
export const registerNameFailed = () => ({
    type: REGISTER_NAME_FAILED
});

export const MNEMONIC_CLOSE = "MNEMONIC_CLOSE";
export const mnemonicClose = () => ({
    type: MNEMONIC_CLOSE
});

export const REGISTERED_NAME_UPDATE_DIALOG = "REGISTERED_NAME_UPDATE_DIALOG";
export const registeredNameUpdateDialog = (changeName) => ({
    type: REGISTERED_NAME_UPDATE_DIALOG,
    payload: {changeName}
});

export const REGISTERED_NAME_UPDATE_DIALOG_CANCEL = "REGISTERED_NAME_UPDATE_DIALOG_CANCEL";
export const registeredNameUpdateDialogCancel = () => ({
    type: REGISTERED_NAME_UPDATE_DIALOG_CANCEL
});

export const REGISTERED_NAME_UPDATE = "REGISTERED_NAME_UPDATE";
export const registeredNameUpdate = (name, generation, mnemonic) => ({
    type: REGISTERED_NAME_UPDATE,
    payload: {name, generation, mnemonic}
});

export const REGISTERED_NAME_UPDATE_SUCCEEDED = "REGISTERED_NAME_UPDATE_SUCCEEDED";
export const registeredNameUpdateSucceeded = () => ({
    type: REGISTERED_NAME_UPDATE_SUCCEEDED
});

export const REGISTERED_NAME_UPDATE_FAILED = "REGISTERED_NAME_UPDATE_FAILED";
export const registeredNameUpdateFailed = () => ({
    type: REGISTERED_NAME_UPDATE_FAILED
});
