// state/registeredname/reducer.js
import {
    REGISTERED_NAME_LOAD,
    REGISTERED_NAME_LOAD_FAILED,
    REGISTERED_NAME_SET,
    REGISTERED_NAME_UNSET
} from "state/registeredname/actions";

const emptyInfo = {
    name: null,
    generation: null,
    operationStatus: null,
    operationStatusUpdated: null,
    operationErrorCode: null,
    operationErrorMessage: null,
    operations: {
        manage: []
    }
};

const initialState = {
    loaded: false,
    loading: false,
    ...emptyInfo
};

export default (state = initialState, action) => {
    switch (action.type) {
        case REGISTERED_NAME_LOAD:
            return {
                ...state,
                loading: true
            };

        case REGISTERED_NAME_LOAD_FAILED:
            return {
                ...state,
                loading: false
            };

        case REGISTERED_NAME_SET:
            return {
                ...state,
                ...emptyInfo,
                ...action.payload.registeredName,
                loading: false,
                loaded: true
            };

        case REGISTERED_NAME_UNSET:
            return {
                ...state,
                ...emptyInfo,
                loading: false,
                loaded: false
            };

        default:
            return state;
    }
}
