// state/registeredname/reducer.js
import {
    MNEMONIC_CLOSE,
    REGISTER_NAME,
    REGISTER_NAME_DIALOG,
    REGISTER_NAME_DIALOG_CANCEL,
    REGISTER_NAME_FAILED,
    REGISTER_NAME_SUCCEEDED,
    REGISTERED_NAME_LOAD,
    REGISTERED_NAME_LOAD_FAILED,
    REGISTERED_NAME_SET,
    REGISTERED_NAME_UNSET
} from "state/registeredname/actions";

const emptyInfo = {
    name: null,
    generation: null,
    operationStatus: null,
    operationStatusUpdated: null,
    operationErrorCode: null,
    operationErrorMessage: null,
    operations: {
        manage: []
    }
};

const initialState = {
    loaded: false,
    loading: false,
    ...emptyInfo,
    showingRegisterDialog: false,
    registering: false,
    mnemonic: null
};

export default (state = initialState, action) => {
    switch (action.type) {
        case REGISTERED_NAME_LOAD:
            return {
                ...state,
                loading: true
            };

        case REGISTERED_NAME_LOAD_FAILED:
            return {
                ...state,
                loading: false
            };

        case REGISTERED_NAME_SET:
            return {
                ...state,
                ...emptyInfo,
                ...action.payload.registeredName,
                loading: false,
                loaded: true
            };

        case REGISTERED_NAME_UNSET:
            return {
                ...state,
                ...emptyInfo,
                loading: false,
                loaded: false
            };

        case REGISTER_NAME_DIALOG:
            return {
                ...state,
                showingRegisterDialog: true
            };

        case REGISTER_NAME_DIALOG_CANCEL:
            return {
                ...state,
                showingRegisterDialog: false
            };

        case REGISTER_NAME:
            return {
                ...state,
                registering: true
            };

        case REGISTER_NAME_SUCCEEDED:
            return {
                ...state,
                showingRegisterDialog: false,
                registering: false,
                mnemonic: action.payload.mnemonic
            };

        case REGISTER_NAME_FAILED:
            return {
                ...state,
                registering: false
            };

        case MNEMONIC_CLOSE:
            return {
                ...state,
                mnemonic: null
            };

        default:
            return state;
    }
}
