// ui/compose/ComposePage.js
import React from 'react';
import { connect } from 'react-redux';
import { connect as connectFormik, Form, withFormik } from 'formik';
import * as yup from 'yup';

import { Page } from "ui/page/Page";
import { Button, Loading, SelectField, TextField } from "ui/control";
import { composePost } from "state/compose/actions";

const SubmitButtonImpl = ({disabled, formik}) => {
    disabled = disabled || formik.values["body"].trim().length === 0;
    return <Button variant="primary" block type="submit" disabled={disabled}>Post</Button>
};

const SubmitButton = connectFormik(SubmitButtonImpl);

const ComposePage = ({loadingFeatures, sourceFormats, beingPosted}) => (
    <Page>
        <h2>New Post <Loading active={loadingFeatures} /></h2>
        <br />
        <Form>
            <SelectField name="bodyFormat" col="col-md-2 pl-0" size="sm" choices={sourceFormats} anyValue />
            <TextField name="body" anyValue autoFocus />
            <SubmitButton disabled={beingPosted} />
            <Loading active={beingPosted} />
        </Form>
    </Page>
);

const composePageLogic = {

    mapPropsToValues(props) {
        return {
            body: props.body || "",
            bodyFormat: props.bodyFormat || "markdown"
        };
    },

    validationSchema: yup.object().shape({
        body: yup.string().trim().required("Must not be empty")
    }),

    handleSubmit(values, formik) {
        formik.props.composePost({
            bodySrc: values.body.trim(),
            bodySrcFormat: values.bodyFormat.trim()
        });
        formik.setSubmitting(false);
    }

};

export default connect(
    state => ({
        loadingFeatures: state.compose.loadingFeatures,
        sourceFormats: state.compose.sourceFormats,
        beingPosted: state.compose.beingPosted
    }),
    {composePost}
)(withFormik(composePageLogic)(ComposePage));
