// ui/compose/ComposePage.js
import React from 'react';
import { connect } from 'react-redux';
import { Form, withFormik } from 'formik';
import * as yup from 'yup';

import { Page } from "ui/page/Page";
import { Button, Loading, TextField } from "ui/control";
import { composePost } from "state/compose/actions";

const ComposePage = ({posting}) => (
    <Page>
        <h2>New Post</h2>
        <br />
        <Form>
            <TextField name="body" anyValue autoFocus />
            <Button variant="primary" block type="submit" disabled={posting}>Post</Button>
            <Loading active={posting} />
        </Form>
    </Page>
);

const composePageLogic = {

    mapPropsToValues(props) {
        return {
            body: props.body || ""
        }
    },

    validationSchema: yup.object().shape({
        body: yup.string().trim().required("Must not be empty")
    }),

    handleSubmit(values, formik) {
        formik.props.composePost({
            bodySrc: values.body.trim()
        });
        formik.setSubmitting(false);
    }

};

export default connect(
    state => ({
        posting: state.compose.posting
    }),
    {composePost}
)(withFormik(composePageLogic)(ComposePage));
