// api/events/actions.js
import immutable from 'object-path-immutable';

export const EVENT_NODE_SUBSCRIBED = "EVENT_NODE_SUBSCRIBED";
export const EVENT_HOME_SUBSCRIBED = "EVENT_HOME_SUBSCRIBED";
export const EVENT_NODE_PROFILE_UPDATED = "EVENT_NODE_PROFILE_UPDATED";
export const EVENT_HOME_NODE_SETTINGS_CHANGED = "EVENT_HOME_NODE_SETTINGS_CHANGED";
export const EVENT_HOME_CLIENT_SETTINGS_CHANGED = "EVENT_HOME_CLIENT_SETTINGS_CHANGED";
export const EVENT_NODE_POSTING_ADDED = "EVENT_NODE_POSTING_ADDED";
export const EVENT_NODE_POSTING_UPDATED = "EVENT_NODE_POSTING_UPDATED";
export const EVENT_NODE_POSTING_DELETED = "EVENT_NODE_POSTING_DELETED";
export const EVENT_NODE_POSTING_RESTORED = "EVENT_NODE_POSTING_RESTORED";
export const EVENT_NODE_REGISTERED_NAME_CHANGED = "EVENT_NODE_REGISTERED_NAME_CHANGED";
export const EVENT_HOME_REGISTERED_NAME_CHANGED = "EVENT_HOME_REGISTERED_NAME_CHANGED";
export const EVENT_NODE_REGISTERED_NAME_OPERATION_STATUS = "EVENT_NODE_REGISTERED_NAME_OPERATION_STATUS";

export const eventAction = (event, prefix) => ({
    type: prefix + event.type,
    payload: immutable.del(event, "type")
});
