// api/events/api.js
import schema from "api/schema";

const baseEvent = (properties = {}) => schema({
    type: "object",
    properties: {
        "type": {
            type: "string"
        },
        ...properties
    },
    additionalProperties: false,
    required: ["type"]
});

const postingEvent = (properties = {}) => baseEvent({
    "id": {
        type: "string"
    },
    "moment": {
        type: "integer"
    },
    ...properties
});

export const EventPacket = schema({
    type: "object",
    properties: {
        "queueStartedAt": {
            type: "integer"
        },
        "ordinal": {
            type: "integer"
        },
        "sentAt": {
            type: "integer"
        },
        "cid": {
            type: "string"
        },
        "event": {
            type: "object",
            properties: {
                "type": {
                    type: "string"
                }
            },
            required: ["type"]
        }
    },
    additionalProperties: false,
    required: ["queueStartedAt", "ordinal", "event"]
});

export const EVENT_SCHEMES = {
    "SUBSCRIBED": baseEvent(),
    "PROFILE_UPDATED": baseEvent(),
    "NODE_SETTINGS_CHANGED": baseEvent(),
    "CLIENT_SETTINGS_CHANGED": baseEvent(),
    "POSTING_ADDED": postingEvent(),
    "POSTING_UPDATED": postingEvent(),
    "POSTING_DELETED": postingEvent(),
    "POSTING_RESTORED": postingEvent()
};
