// api/events/api.js
import schema from "api/schema";

const ProfileUpdatedEvent = {
    type: "object",
    properties: {
        "type": {
            type: "string",
            const: "PROFILE_UPDATED"
        }
    },
    additionalProperties: false,
    required: ["type"]
};

const NodeSettingsChangedEvent = {
    type: "object",
    properties: {
        "type": {
            type: "string",
            const: "NODE_SETTINGS_CHANGED"
        }
    },
    additionalProperties: false,
    required: ["type"]
};

const ClientSettingsChangedEvent = {
    type: "object",
    properties: {
        "type": {
            type: "string",
            const: "CLIENT_SETTINGS_CHANGED"
        }
    },
    additionalProperties: false,
    required: ["type"]
};

const Event = {
    anyOf: [
        ProfileUpdatedEvent,
        NodeSettingsChangedEvent,
        ClientSettingsChangedEvent
    ]
};

export const EventPacket = schema({
    type: "object",
    properties: {
        "queueStartedAt": {
            type: "integer"
        },
        "ordinal": {
            type: "integer"
        },
        "sentAt": {
            type: "integer"
        },
        "cid": {
            type: "string"
        },
        "event": Event
    },
    additionalProperties: false,
    required: ["queueStartedAt", "ordinal", "event"]
});
