// mainmenu/owner/OwnerName.js
import React from 'react';
import PropType from 'prop-types';
import { connect } from 'react-redux';
import cx from 'classnames';

import "mainmenu/owner/OwnerName.css";
import { ownerLoad } from "mainmenu/owner/actions";

const OwnerNameGeneration = ({generation, latest}) => (
    latest || <span className="generation">{generation}</span>
);

OwnerNameGeneration.propTypes = {
    generation: PropType.number,
    latest: PropType.bool
};

class OwnerName extends React.PureComponent {

    static propTypes = {
        name: PropType.string,
        generation: PropType.number,
        latest: PropType.bool,
        verified: PropType.bool,
        correct: PropType.bool
    };

    componentDidMount() {
        this.props.ownerLoad();
    }

    render() {
        const { name, generation, latest, verified, correct } = this.props;

        if (name) {
            const klass = cx([
                "owner-name", {
                    "correct": verified && correct,
                    "incorrect": verified && !correct
                }
            ]);
            return (
                <span className={klass}>
                    {name}
                    <OwnerNameGeneration generation={generation} latest={latest} />
                </span>
            );
        } else {
            return null;
        }
    }

}

export default connect(
    (state) => state.owner,
    { ownerLoad }
)(OwnerName);
