// profile/edit/ProfileEditor.js
import React from 'react';
import PropType from 'prop-types';
import { connect } from 'react-redux';
import { Form, withFormik } from 'formik';

import { Button, InputField, Loading } from "control";
import { profileEditCancel, profileUpdate } from "profile/actions";
import "profile/edit/ProfileEditor.css";

const ProfileLine = ({title, name, autoFocus = false}) => (
    <>
        <dt className="col-sm-2">{title}</dt>
        <dd className="col-sm-10"><InputField name={name} autoFocus={autoFocus} /></dd>
    </>
);

ProfileLine.propTypes = {
    title: PropType.string,
    name: PropType.string,
    autoFocus: PropType.bool
};

class ProfileEditor extends React.PureComponent {

    static propTypes = {
        loading: PropType.bool,
        updating: PropType.bool,
        fullName: PropType.string,
        gender: PropType.string
    };

    render() {
        const {loading, updating, profileEditCancel} = this.props;

        return (
            <>
                <h2>
                    Profile <Loading active={loading} />
                </h2>
                <Form>
                    <dl className="row">
                        <ProfileLine title="Full name" name="fullName" autoFocus />
                        <ProfileLine title="Gender" name="gender" />
                    </dl>
                    <div className="editor-footer">
                        <Loading active={updating} />
                        <Button variant="secondary" onClick={profileEditCancel} disabled={updating}>Cancel</Button>
                        <Button variant="primary" type="submit" disabled={updating}>Update</Button>
                    </div>
                </Form>
            </>
        );
    }

}

const profileEditorLogic = {

    mapPropsToValues(props) {
        return {
            fullName: props.fullName || "",
            gender: props.gender || ""
        }
    },

    handleSubmit(values, formik) {
        formik.props.profileUpdate(values);
        formik.setSubmitting(false);
    }

};

export default connect(
    (state) => ({
        loading: state.profile.loading,
        updating: state.profile.updating,
        fullName: state.profile.fullName,
        gender: state.profile.gender
    }),
    {profileEditCancel, profileUpdate}
)(withFormik(profileEditorLogic)(ProfileEditor));
