// state/reactionsdialog/actions.js
export const OPEN_REACTIONS_DIALOG = "OPEN_REACTIONS_DIALOG";
export const openReactionsDialog = (postingId, negative) => ({
    type: OPEN_REACTIONS_DIALOG,
    payload: {postingId, negative}
});

export const CLOSE_REACTIONS_DIALOG = "CLOSE_REACTIONS_DIALOG";
export const closeReactionsDialog = () => ({
    type: CLOSE_REACTIONS_DIALOG
});

export const REACTIONS_DIALOG_PAST_REACTIONS_LOAD = "REACTIONS_DIALOG_PAST_REACTIONS_LOAD";
export const reactionsDialogPastReactionsLoad = () => ({
    type: REACTIONS_DIALOG_PAST_REACTIONS_LOAD
});

export const REACTIONS_DIALOG_PAST_REACTIONS_LOADED = "REACTIONS_DIALOG_PAST_REACTIONS_LOADED";
export const reactionsDialogPastReactionsLoaded = (reactions, postingId, negative, before, after, total) => ({
    type: REACTIONS_DIALOG_PAST_REACTIONS_LOADED,
    payload: {reactions, postingId, negative, before, after, total}
});

export const REACTIONS_DIALOG_PAST_REACTIONS_LOAD_FAILED = "REACTIONS_DIALOG_PAST_REACTIONS_LOAD_FAILED";
export const reactionsDialogPastReactionsLoadFailed = () => ({
    type: REACTIONS_DIALOG_PAST_REACTIONS_LOAD_FAILED
});

export const REACTIONS_DIALOG_UNSET = "REACTIONS_DIALOG_UNSET";
export const reactionsDialogUnset = () => ({
    type: REACTIONS_DIALOG_UNSET
});
