// state/reactionsdialog/reducer.js
import { CLOSE_REACTIONS_DIALOG, OPEN_REACTIONS_DIALOG } from "state/reactionsdialog/actions";

const initialState = {
    show: false,
    postingId: null,
    negative: false,
    reactions: {
        loading: false,
        total: 0,
        after: Number.MAX_SAFE_INTEGER,
        items: []
    },
    totals: {
        loading: false,
        loaded: false,
        total: 0,
        emojis: []
    }
};

export default (state = initialState, action) => {
    switch (action.type) {
        case OPEN_REACTIONS_DIALOG:
            if (state.postingId === action.payload.postingId && state.negative === action.payload.negative) {
                return {
                    ...state,
                    show: true
                }
            }
            return {
                ...initialState,
                show: true,
                postingId: action.payload.postingId,
                negative: action.payload.negative
            };

        case CLOSE_REACTIONS_DIALOG:
            return {
                ...state,
                show: false
            };

        default:
            return state;
    }
}
