// ui/profile/manage/RegisteredNameUpdateDialog.js
import React from 'react';
import { connect } from 'react-redux';
import { Form, withFormik } from 'formik';
import * as yup from 'yup';

import { Button, InputField, Loading, ModalDialog, NumberField } from "ui/control";
import { registeredNameUpdate, registeredNameUpdateDialogCancel } from "state/registeredname/actions";
import * as Rules from "api/naming/rules";
import { range } from "util/util";
import "./RegisteredNameUpdateDialog.css";

const Column = ({start, end, autoFocus = false}) => (
    <div className="col-sm-4">
        <ol start={start + 1}>
            {range(end - start).map(index => (
                <li key={index}>
                    <InputField name={`mnemonic[${start + index}]`} className="mnemonic-input"
                                autoFocus={index === 0 && autoFocus} noFeedback />
                </li>
            ))}
        </ol>
    </div>
);

class RegisteredNameUpdateDialog extends React.Component {

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (this.props.show !== prevProps.show) { // Mnemonic must be cleared immediately after hiding the dialog
            this.props.resetForm();
        }
    }

    render() {
        const {show, showChangeName, updating, registeredNameUpdateDialogCancel} = this.props;

        if (!show) {
            return null;
        }

        return (
            <ModalDialog title={showChangeName ? "Change Registered Name" : "Prolong RegisteredName"} size="lg"
                         onClose={registeredNameUpdateDialogCancel}>
                <Form>
                    <div className="modal-body">
                        {showChangeName &&
                        <div className="row">
                            <div className="col-sm-4">
                                <InputField name="name" title="Name" autoFocus/>
                            </div>
                            <div className="col-sm-2">
                                <NumberField name="generation" title="Generation" min={0}/>
                            </div>
                        </div>}
                        <h5 className="mnemonic-title">Secret Words</h5>
                        <div className="row">
                            <Column start={0} end={8} autoFocus={!showChangeName} />
                            <Column start={8} end={16} />
                            <Column start={16} end={24} />
                        </div>
                    </div>
                    <div className="modal-footer">
                        <Loading active={updating}/>
                        <Button variant="secondary" onClick={registeredNameUpdateDialogCancel}
                                disabled={updating}>Cancel</Button>
                        <Button variant="primary" type="submit" disabled={updating}>
                            {showChangeName ? "Change" : "Prolong"}
                        </Button>
                    </div>
                </Form>
            </ModalDialog>
        );
    }

}

const registeredNameUpdateDialogLogic = {

    mapPropsToValues(props) {
        return {
            name: props.name ? props.name : "",
            generation: props.generation ? props.generation : 0,
            mnemonic: Array(24).fill("")
        }
    },

    validationSchema(props) {
        const mnemonic = yup.array().transform((value, originalValue) => (
                Array(24).fill("").map((v, i) => (value[i] ? value[i] : v))
            )).of(yup.string().trim().required("Must not be empty")
                    .matches(/^[A-Za-z]*$/, "Must be a single English word"));
        return props.showChangeName
            ? yup.object().shape({
                name: yup.string().trim().required("Must not be empty").max(Rules.NAME_MAX_LENGTH)
                    .test("is-allowed", "Name is not allowed", Rules.isNameValid),
                generation: yup.number().min(0, "Must be 0 or greater"),
                mnemonic
            })
            : yup.object().shape({
                mnemonic
            })
    },

    handleSubmit(values, formik) {
        formik.props.registeredNameUpdate(values.name.trim(), values.generation,
            values.mnemonic.map(v => v.trim().toLowerCase()));
        formik.setSubmitting(false);
    }

};

export default connect(
    state => ({
        show: state.registeredName.showingUpdateDialog,
        showChangeName: state.registeredName.showingChangeName,
        updating: state.registeredName.updating,
        name: state.registeredName.name,
        generation: state.registeredName.generation
    }),
    { registeredNameUpdateDialogCancel, registeredNameUpdate }
)(withFormik(registeredNameUpdateDialogLogic)(RegisteredNameUpdateDialog));
