// state/settings/actions.js
export const SETTINGS_NODE_VALUES_LOAD = "SETTINGS_NODE_VALUES_LOAD";
export const settingsNodeValuesLoad = () => ({
    type: SETTINGS_NODE_VALUES_LOAD
});

export const SETTINGS_NODE_VALUES_LOADED = "SETTINGS_NODE_VALUES_LOADED";
export const settingsNodeValuesLoaded = (settings) => ({
    type: SETTINGS_NODE_VALUES_LOADED,
    payload: {settings}
});

export const SETTINGS_NODE_VALUES_LOAD_FAILED = "SETTINGS_NODE_VALUES_LOAD_FAILED";
export const settingsNodeValuesLoadFailed = () => ({
    type: SETTINGS_NODE_VALUES_LOAD_FAILED
});

export const SETTINGS_NODE_META_LOAD = "SETTINGS_NODE_META_LOAD";
export const settingsNodeMetaLoad = () => ({
    type: SETTINGS_NODE_META_LOAD
});

export const SETTINGS_NODE_META_LOADED = "SETTINGS_NODE_META_LOADED";
export const settingsNodeMetaLoaded = (meta) => ({
    type: SETTINGS_NODE_META_LOADED,
    payload: {meta}
});

export const SETTINGS_NODE_META_LOAD_FAILED = "SETTINGS_NODE_META_LOAD_FAILED";
export const settingsNodeMetaLoadFailed = () => ({
    type: SETTINGS_NODE_META_LOAD_FAILED
});

export const SETTINGS_UPDATE = "SETTINGS_UPDATE";
export const settingsUpdate = (settings, onSuccess) => ({
    type: SETTINGS_UPDATE,
    payload: {settings, onSuccess}
});

export const SETTINGS_UPDATE_SUCCEEDED = "SETTINGS_UPDATE_SUCCEEDED";
export const settingsUpdateSucceeded = (settings, onSuccess) => ({
    type: SETTINGS_UPDATE_SUCCEEDED,
    payload: {settings, onSuccess}
});

export const SETTINGS_UPDATE_FAILED = "SETTINGS_UPDATE_FAILED";
export const settingsUpdateFailed = () => ({
    type: SETTINGS_UPDATE_FAILED
});
