// state/owner/reducer.js
import moment from 'moment';

import { OWNER_SET, OWNER_VERIFIED } from "state/owner/actions";

const initialState = {
    name: null,
    latest: false,
    verified: false,
    correct: false,
    deadline: null,
    verifiedAt: 0
};

export default (state = initialState, action) => {
    switch (action.type) {
        case OWNER_SET:
            if (state.name !== action.payload.name) {
                return {
                    ...state,
                    ...action.payload,
                    verified: false,
                    verifiedAt: 0
                };
            } else {
                return state;
            }

        case OWNER_VERIFIED:
            if (state.name === action.payload.name) {
                return {
                    ...state,
                    ...action.payload,
                    verified: true,
                    verifiedAt: moment().unix()
                };
            }
            return state;

        default:
            return state;
    }
}
