// ui/profile/edit/ProfileEditor.js
import React from 'react';
import PropType from 'prop-types';
import { connect } from 'react-redux';
import { Form, withFormik } from 'formik';

import { Button, InputField, Loading } from "ui/control";
import { profileEditCancel, profileUpdate } from "state/profile/actions";
import "./ProfileEditor.css";

const ProfileLine = ({title, children}) => (
    <>
        <dt className="col-sm-2">{title}</dt>
        <dd className="col-sm-10">{children}</dd>
    </>
);

ProfileLine.propTypes = {
    title: PropType.string
};

class ProfileEditor extends React.PureComponent {

    static propTypes = {
        loading: PropType.bool,
        loaded: PropType.bool,
        updating: PropType.bool,
        fullName: PropType.string,
        gender: PropType.string,
        email: PropType.string
    };

    render() {
        const {loading, updating, profileEditCancel} = this.props;

        return (
            <>
                <h2>
                    Edit Profile <Loading active={loading} />
                </h2>
                <Form>
                    <dl className="row">
                        <ProfileLine title="Full name">
                            <InputField name="fullName" autoFocus />
                        </ProfileLine>
                        <ProfileLine title="Gender">
                            <InputField name="gender" data={["Male", "Female"]} />
                        </ProfileLine>
                        <ProfileLine title="E-Mail">
                            <InputField name="email" />
                        </ProfileLine>
                    </dl>
                    <div className="editor-footer">
                        <Loading active={updating} />
                        <Button variant="secondary" onClick={profileEditCancel} disabled={updating}>Cancel</Button>
                        <Button variant="primary" type="submit" disabled={updating}>Update</Button>
                    </div>
                </Form>
            </>
        );
    }

}

const profileEditorLogic = {

    mapPropsToValues(props) {
        return {
            fullName: props.fullName || "",
            gender: props.gender || "",
            email: props.email || ""
        }
    },

    handleSubmit(values, formik) {
        formik.props.profileUpdate(values);
        formik.setSubmitting(false);
    }

};

export default connect(
    (state) => ({
        loading: state.profile.loading,
        loaded: state.profile.loaded,
        updating: state.profile.updating,
        fullName: state.profile.fullName,
        gender: state.profile.gender,
        email: state.profile.email
    }),
    {profileEditCancel, profileUpdate}
)(withFormik(profileEditorLogic)(ProfileEditor));
