// ui/timeline/TimelinePage.js
import React from 'react';
import { connect } from 'react-redux';

import { Page } from "ui/page/Page";
import { Loading } from "ui/control";
import Posting from "ui/timeline/Posting";
import { timelineFutureSliceLoad, timelinePastSliceLoad, timelineScrolled } from "state/timeline/actions";
import { isAtTimelinePage } from "state/navigation/selectors";

class TimelinePage extends React.Component {

    constructor(props) {
        super(props);
        this.prevAt = Number.MAX_SAFE_INTEGER;

        this.onScroll = this.onScroll.bind(this);
        this.updateOnScroll();
        this.observeSentinelFuture = this.observeSentinelFuture.bind(this);
        this.loadFuture = this.loadFuture.bind(this);
        this.observerFuture = new IntersectionObserver(this.loadFuture, {rootMargin: "25% 0% 0% 0%"});
        this.observeSentinelPast = this.observeSentinelPast.bind(this);
        this.loadPast = this.loadPast.bind(this);
        this.observerPast = new IntersectionObserver(this.loadPast, {rootMargin: "0% 0% 25% 0%"});
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (this.props.visible !== prevProps.visible) {
            this.updateOnScroll();
        }
        if (this.props.visible) {
            if (this.props.anchor !== prevProps.anchor) {
                this.scrollTo(this.props.anchor);
            }
            const topmost = TimelinePage.getTopmostMoment();
            if (topmost !== this.prevAt) {
                this.scrollTo(this.prevAt);
            }
        }
    }

    updateOnScroll() {
        window.onscroll = this.props.visible ? this.onScroll : null;
    }

    onScroll() {
        const at = TimelinePage.getTopmostMoment();
        if (at !== this.prevAt) {
            this.props.timelineScrolled(at);
        }
        this.prevAt = at;

    }

    static getTopmostMoment() {
        const postings = document.getElementsByClassName("posting");
        for (let i = 0; i < postings.length; i++) {
            if (postings.item(i).getBoundingClientRect().top >= 0) {
                return parseInt(postings.item(i).dataset.moment);
            }
        }
        return Number.MAX_SAFE_INTEGER;
    }

    static getPostingAt(moment) {
        const postings = document.getElementsByClassName("posting");
        for (let i = 0; i < postings.length; i++) {
            if (postings.item(i).dataset.moment <= moment) {
                return postings.item(i);
            }
        }
        return null;
    }

    scrollTo(moment) {
        const posting = TimelinePage.getPostingAt(moment);
        if (posting != null) {
            posting.scrollIntoView(true);
            const mainMenu = document.getElementById("main-menu");
            if (mainMenu != null) {
                window.scrollBy(0, -mainMenu.getBoundingClientRect().height - 10);
            }
        }
        this.prevAt = moment;
    }

    observeSentinelFuture(sentinel) {
        if (sentinel == null) {
            this.observerFuture.disconnect();
        } else {
            this.observerFuture.observe(sentinel);
        }
    }

    loadFuture() {
        if (this.props.loadingFuture || this.props.before >= Number.MAX_SAFE_INTEGER) {
            return;
        }
        this.props.timelineFutureSliceLoad();
    }

    observeSentinelPast(sentinel) {
        if (sentinel == null) {
            this.observerPast.disconnect();
        } else {
            this.observerPast.observe(sentinel);
        }
    }

    loadPast() {
        if (this.props.loadingPast || this.props.after <= Number.MIN_SAFE_INTEGER) {
            return;
        }
        this.props.timelinePastSliceLoad();
    }

    render() {
        const {loadingFuture, loadingPast, postings} = this.props;
        return (
            <Page>
                <h2>Timeline</h2>
                <Loading active={loadingFuture} />
                <div ref={this.observeSentinelFuture} />
                {postings.map(posting => <Posting key={posting.moment} posting={posting} />)}
                <div ref={this.observeSentinelPast} />
                <Loading active={loadingPast} />
            </Page>
        );
    }
}

export default connect(
    state => ({
        visible: isAtTimelinePage(state),
        loadingFuture: state.timeline.loadingFuture,
        loadingPast: state.timeline.loadingPast,
        before: state.timeline.before,
        after: state.timeline.after,
        postings: state.timeline.postings,
        anchor: state.timeline.anchor
    }),
    { timelineFutureSliceLoad, timelinePastSliceLoad, timelineScrolled }
)(TimelinePage);
