// ui/timeline/TimelinePageHeader.js
import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { DateTimePicker } from 'react-widgets';

import { getTimelineAtTimestamp } from "state/timeline/selectors";
import { goToTimeline } from "state/navigation/actions";
import "./TimelinePageHeader.css";

class TimelinePageHeader extends React.PureComponent {

    constructor(props) {
        super(props);

        this.state = {active: false};
        this.activate = this.activate.bind(this);
        this.goToTimestamp = this.goToTimestamp.bind(this);
    }


    componentDidUpdate(prevProps, prevState, snapshot) {
        if (this.props.timestamp !== prevProps.timestamp) {
            this.setState({active: false});
        }
    }

    activate() {
        this.setState({active: true});
    }

    goToTimestamp(value) {
        this.props.goToTimeline(moment(value).endOf('day').unix() * 100);
    }

    render() {
        const {empty = false, timestamp} = this.props;
        const {active} = this.state;

        return (
            <div id="timeline-header">
                <h2>Timeline</h2>
                {!empty &&
                    <div id="timeline-calendar-btn">
                        {!active &&
                            <button className="btn btn-outline-info btn-sm" onClick={this.activate}>Calendar</button>
                        }
                        {active &&
                            <DateTimePicker format="dd-MM-yyyy" value={moment.unix(timestamp).toDate()} time={false}
                                            onChange={this.goToTimestamp}/>
                        }
                    </div>
                }
            </div>
        );
    }

}

export default connect(
    state => ({
        timestamp: getTimelineAtTimestamp(state)
    }),
    { goToTimeline }
)(TimelinePageHeader);
