// naming/api.js
import schema from 'schm';

export const Result = schema({
    jsonrpc: {
        type: String,
        required: true
    },
    result: {
        type: Object,
        required: true
    },
    id: {
        type: Number,
        required: true
    }
});

export const ErrorResult = schema({
    jsonrpc: {
        type: String,
        required: true
    },
    error: {
        code: {
            type: Number,
            required: true
        },
        message: {
            type: String,
            required: true
        }
    },
    id: {
        type: Number,
        required: true
    }
});

export const RegisteredNameInfo = schema({
    name: {
        type: String,
        required: true
    },
    generation: {
        type: Number,
        required: true,
        min: 0
    },
    latest: {
        type: Boolean,
        required: true
    },
    updatingKey: {
        type: String,
        required: true
    },
    nodeUri: String,
    deadline: {
        type: Number,
        required: true
    },
    signingKey: String,
    validFrom: Number
});
