// naming/sagas.js
import { select, call, apply } from 'redux-saga/effects';

import { formatSchemaErrors, NamingError } from "error";
import * as NamingApi from "naming/api";

let callId = 1;

function* callNaming({method, params, schema = null, notNull = false}) {
    const exception = (e, details = null) => new NamingError(method, e, details);
    const location = yield select(state => state.naming.location);
    let response;
    try {
        response = yield call(fetch, location, {
            method: "POST",
            headers: {
                "Accept": "application/json",
                "Content-Type": "application/json"
            },
            body: JSON.stringify({
                jsonrpc: "2.0",
                id: callId++,
                method,
                params
            })
        });
    } catch (e) {
        throw exception(e);
    }
    if (!response.ok) {
        throw exception("Server returned error status");
    }
    let data;
    try {
        data = yield apply(response, response.json);
    } catch (e) {
        throw exception("Server returned empty result");
    }
    if (response.error) {
        if (!NamingApi.ErrorResult(data)) {
            throw exception("Error response format incorrect", formatSchemaErrors(NamingApi.ErrorResult.errors));
        }
        throw exception(data.error.message);
    }
    if (!NamingApi.Result(data)) {
        throw exception("Response format incorrect", formatSchemaErrors(NamingApi.Result.errors));
    }
    if (data.result == null) {
        if (!notNull) {
            return null;
        } else {
            throw exception("Return value is null");
        }
    }
    if (schema && !schema(data.result)) {
        throw exception("Return value format incorrect: ", formatSchemaErrors(schema.errors));
    }
    return data.result;
}

export function* getCurrent(name, generation) {
    return yield call(callNaming,
        {method: "getCurrent", params: [name, generation], schema: NamingApi.RegisteredNameInfo});
}
