// profile/profileeditor/ProfileEditor.js
import React from 'react';
import PropType from 'prop-types';
import { connect } from 'react-redux';
import { Form, withFormik } from 'formik';

import { Button, InputField, Loading } from "control";
import { profileEditCancel } from "profile/actions";
import "profile/profileeditor/ProfileEditor.css";

const ProfileLine = ({title, name, autoFocus = false}) => (
    <>
        <dt className="col-sm-2">{title}</dt>
        <dd className="col-sm-10"><InputField name={name} autoFocus={autoFocus} /></dd>
    </>
);

ProfileLine.propTypes = {
    title: PropType.string,
    name: PropType.string,
    autoFocus: PropType.bool
};

class ProfileEditor extends React.PureComponent {

    static propTypes = {
        loading: PropType.bool,
        fullName: PropType.string,
        gender: PropType.string
    };

    render() {
        const {loading, profileEditCancel} = this.props;

        return (
            <>
                <h2>
                    Profile <Loading active={loading} />
                </h2>
                <Form>
                    <dl className="row">
                        <ProfileLine title="Full name" name="fullName" autoFocus />
                        <ProfileLine title="Gender" name="gender" />
                    </dl>
                    <div className="editor-footer">
                        <Button variant="secondary" onClick={profileEditCancel}>Cancel</Button>
                        <Button variant="primary" type="submit">Update</Button>
                    </div>
                </Form>
            </>
        );
    }

}

const profileEditorLogic = {

    mapPropsToValues(props) {
        return {
            fullName: props.fullName || "",
            gender: props.gender || ""
        }
    }

};

export default connect(
    (state) => ({
        ...state.profile.profileEditor,
        loading: state.profile.loading
    }),
    {profileEditCancel}
)(withFormik(profileEditorLogic)(ProfileEditor));
