// ui/control/CheckboxField.js
import React from 'react';
import PropType from 'prop-types';
import { Field } from 'formik';

import "./CheckboxField.css";

export class CheckboxField extends React.PureComponent {

    static propTypes = {
        name: PropType.string,
        title: PropType.string,
        autoFocus: PropType.bool
    };

    constructor(props) {
        super(props);
        this.inputDom = null;
    }

    componentDidMount() {
        if (this.props.autoFocus && this.inputDom) {
            this.inputDom.focus();
        }
    }

    render() {
        const {name, title} = this.props;

        return (
            <div className="form-group">
                <div className="form-check">
                    <Field name={name}>
                        {({field, form}) => {
                            const touched = form.touched[field.name];
                            const error = form.errors[field.name];
                            return (
                                <>
                                    <input
                                        {...field}
                                        id={name}
                                        type="checkbox"
                                        className="form-check-input"
                                        ref={dom => {this.inputDom = dom}}
                                    />
                                    {touched && error && <div className="invalid-feedback">{error}</div>}
                                </>
                            );
                        }}
                    </Field>
                    {title && <label className="form-check-label checkbox-field-label" htmlFor={name}>{title}</label>}
                </div>
            </div>
        );
    }

}
