// ui/control/CheckboxField.js
import React from 'react';
import PropType from 'prop-types';
import { Field } from 'formik';

import { FormGroup } from "ui/control";

import "./CheckboxField.css";

export class CheckboxField extends React.PureComponent {

    static propTypes = {
        name: PropType.string,
        title: PropType.string,
        autoFocus: PropType.bool
    };

    constructor(props) {
        super(props);
        this.inputDom = null;
    }

    componentDidMount() {
        if (this.props.autoFocus && this.inputDom) {
            this.inputDom.focus();
        }
    }

    render() {
        const {name, title} = this.props;

        return (
            <FormGroup title={title} name={name} checkbox={true}>
                <Field name={name}>
                    {({field, form}) => {
                        const touched = form.touched[field.name];
                        const error = form.errors[field.name];
                        return (
                            <>
                                <input
                                    {...field}
                                    checked={field.value}
                                    id={name}
                                    type="checkbox"
                                    className="form-check-input"
                                    ref={dom => {this.inputDom = dom}}
                                />
                                {touched && error && <div className="invalid-feedback">{error}</div>}
                            </>
                        );
                    }}
                </Field>
            </FormGroup>
        );
    }

}
