// ui/control/ComboboxField.js
import React from 'react';
import PropType from 'prop-types';
import { Field } from 'formik';
import { Combobox } from 'react-widgets';

export class ComboboxField extends React.PureComponent {

    static propTypes = {
        name: PropType.string,
        title: PropType.string,
        autoFocus: PropType.bool,
        data: PropType.array
    };

    render() {
        const {name, title, autoFocus, data} = this.props;

        return (
            <div className="form-group">
                {title && <label htmlFor={name}>{title}</label>}
                <Field name={name}>
                    {({field, form}) => {
                        const touched = form.touched[field.name];
                        const error = form.errors[field.name];
                        return (
                            <>
                                <Combobox
                                    name={field.name}
                                    value={field.value}
                                    onChange={v => form.setFieldValue(field.name, v)}
                                    autoFocus={autoFocus}
                                    data={data}
                                />
                                {touched && error && <div className="invalid-feedback">{error}</div>}
                            </>
                        );
                    }}
                </Field>
            </div>
        );
    }

}
