// ui/control/DurationField.js
import React from 'react';
import PropType from 'prop-types';
import { NumberPicker } from 'react-widgets';
import { Field } from 'formik';
import cx from 'classnames';
import selectn from 'selectn';

import { Column, FormGroup } from "ui/control";
import "./DurationField.css";

export class DurationField extends React.PureComponent {

    static propTypes = {
        name: PropType.string,
        title: PropType.string,
        horizontal: PropType.bool,
        groupClassName: PropType.string,
        labelClassName: PropType.string,
        col: PropType.string,
        autoFocus: PropType.bool,
        noFeedback: PropType.bool
    };

    constructor(props) {
        super(props);

        this.onChange = this.onChange.bind(this);
    }

    static parseValue(v) {
        if (v == null || v.length === 0) {
            return {unit: "s", duration: 0};
        }
        return {
            unit: v.charAt(v.length - 1),
            duration: parseInt(v.substring(0, v.length - 1))
        }
    }

    onChange(form, fieldName, newUnit, newDuration) {
        let {unit, duration} = DurationField.parseValue(form.values[fieldName]);
        unit = newUnit != null ? newUnit : unit;
        duration = newDuration != null ? newDuration : duration;
        form.setFieldValue(fieldName, `${duration}${unit}`);
    }

    render() {
        const {name, title, horizontal = false, groupClassName, labelClassName, col, noFeedback = false,
               autoFocus} = this.props;

        return (
            <FormGroup title={title} name={name + "_duration"} horizontal={horizontal} labelClassName={labelClassName}
                       groupClassName={groupClassName}>
                {/* <label> is not functional, because NumberPicker doesn't allow to set id */}
                <Field name={name}>
                    {({field, form}) => {
                        const {unit, duration} = DurationField.parseValue(field.value);
                        const touched = selectn(field.name, form.touched);
                        const error = selectn(field.name, form.errors);
                        return (
                            <Column className={col}>
                                <div className="duration-control">
                                    <NumberPicker
                                        name={field.name + "_duration"}
                                        value={duration}
                                        onChange={v => this.onChange(form, field.name, null, v)}
                                        autoFocus={autoFocus}
                                        min={0}
                                    />
                                    <select name={field.name + "_unit"} value={unit}
                                            onChange={e => this.onChange(form, field.name, e.target.value, null)}>
                                        <option key="s" value="s">seconds</option>
                                        <option key="m" value="m">minutes</option>
                                        <option key="h" value="h">hours</option>
                                        <option key="d" value="d">days</option>
                                    </select>
                                </div>
                                {!noFeedback && touched && error && <div className="invalid-feedback">{error}</div>}
                            </Column>
                        );
                    }}
                </Field>
            </FormGroup>
        );
    }

}
