// ui/control/InputField.js
import React from 'react';
import PropType from 'prop-types';
import { Field } from 'formik';
import cx from 'classnames';
import selectn from 'selectn';

import { Column, FormGroup } from "ui/control";

export class InputField extends React.PureComponent {

    static propTypes = {
        name: PropType.string,
        title: PropType.string,
        horizontal: PropType.bool,
        groupClassName: PropType.string,
        labelClassName: PropType.string,
        col: PropType.string,
        autoFocus: PropType.bool,
        anyValue: PropType.bool,
        className: PropType.string,
        autoComplete: PropType.string,
        noFeedback: PropType.bool
    };

    constructor(props) {
        super(props);
        this.inputDom = null;
    }

    componentDidMount() {
        if (this.props.autoFocus && this.inputDom) {
            this.inputDom.focus();
        }
    }

    render() {
        const {name, title, horizontal = false, groupClassName, labelClassName, col, anyValue, className, autoComplete,
               noFeedback = false} = this.props;

        return (
            <FormGroup title={title} name={name} horizontal={horizontal} labelClassName={labelClassName}
                       groupClassName={groupClassName}>
                <Field name={name}>
                    {({field, form}) => {
                        const touched = selectn(field.name, form.touched);
                        const error = selectn(field.name, form.errors);
                        return (
                            <Column className={col}>
                                <input
                                    {...field}
                                    id={name}
                                    type={!name.toLowerCase().includes("password") ? "text" : "password"}
                                    className={cx(
                                        "form-control", {
                                            "is-valid": !anyValue && touched && !error,
                                            "is-invalid": !anyValue && touched && error,
                                            [className]: !!className
                                        })}
                                    autoComplete={autoComplete}
                                    ref={dom => this.inputDom = dom}
                                />
                                {!noFeedback && touched && error && <div className="invalid-feedback">{error}</div>}
                            </Column>
                        );
                    }}
                </Field>
            </FormGroup>
        );
    }

}
