// ui/control/NumberField.js
import React from 'react';
import PropType from 'prop-types';
import { Field } from 'formik';
import { NumberPicker } from 'react-widgets';

export const NumberField = ({name, title, autoFocus, format, min, max}) => (
    <div className="form-group">
        {/* <label> is not functional, because NumberPicker doesn't allow to set id */}
        {title && <label htmlFor={name}>{title}</label>}
        <Field name={name}>
            {({field, form}) => {
                const touched = form.touched[field.name];
                const error = form.errors[field.name];
                return (
                    <>
                        <NumberPicker
                            name={field.name}
                            value={field.value}
                            onChange={v => form.setFieldValue(field.name, v)}
                            autoFocus={autoFocus}
                            format={format}
                            min={min}
                            max={max}
                        />
                        {touched && error && <div className="invalid-feedback">{error}</div>}
                    </>
                );
            }}
        </Field>
    </div>
);

NumberField.propTypes = {
    name: PropType.string,
    title: PropType.string,
    autoFocus: PropType.bool,
    format: PropType.string,
    min: PropType.number,
    max: PropType.number
};
