// ui/control/NumberField.js
import React from 'react';
import PropType from 'prop-types';
import { Field } from 'formik';
import { NumberPicker } from 'react-widgets';

import { FormGroup } from "ui/control";
import "./NumberField.css";

export const NumberField = ({name, title, autoFocus, format, min, max, initialValue, defaultValue}) => (
    <Field name={name}>
        {({field, form}) => {
            const touched = form.touched[field.name];
            const error = form.errors[field.name];
            return (
                <FormGroup
                    title={title}
                    name={name}
                    undo={field.value !== initialValue}
                    reset={field.value !== defaultValue}
                    onUndo={() => form.setFieldValue(field.name, initialValue)}
                    onReset={() => form.setFieldValue(field.name, defaultValue)}
                >
                    <>
                        {/* <label> is not functional, because NumberPicker doesn't allow to set id */}
                        <div className="number-control">
                            <NumberPicker
                                name={field.name}
                                value={field.value}
                                onChange={v => form.setFieldValue(field.name, v)}
                                autoFocus={autoFocus}
                                format={format}
                                min={min}
                                max={max}
                            />
                        </div>
                        {touched && error && <div className="invalid-feedback">{error}</div>}
                    </>
                </FormGroup>
            );
        }}
    </Field>
);

NumberField.propTypes = {
    name: PropType.string,
    title: PropType.string,
    autoFocus: PropType.bool,
    format: PropType.string,
    min: PropType.number,
    max: PropType.number,
    initialValue: PropType.number,
    defaultValue: PropType.number
};
