// ui/control/Popover.js
import React from 'react';
import * as ReactDOM from 'react-dom';
import cx from 'classnames';
import { Manager, Popper, Reference } from 'react-popper';

export class Popover extends React.PureComponent {

    constructor(props) {
        super(props);
        this.state = {visible: false};

        this.toggle = this.toggle.bind(this);
        this.documentClick = this.documentClick.bind(this);
    }

    toggle() {
        if (!this.state.visible) {
            this.show();
        } else {
            this.hide();
        }
    }

    documentClick(event) {
        for (let element of document.querySelectorAll(".popover.show").values()) {
            const r = element.getBoundingClientRect();
            if (r.left <= event.clientX && r.right >= event.clientX
                && r.top <= event.clientY && r.bottom >= event.clientY) {
                return;
            }
        }
        this.hide();
    }

    show() {
        this.setState({visible: true});
        document.addEventListener("click", this.documentClick);
    }

    hide() {
        this.setState({visible: false});
        document.removeEventListener("click", this.documentClick);
    }

    render() {
        const pointerEvents = {pointerEvents: this.state.visible ? "auto" : "none"};
        return (
            <Manager>
                <Reference>
                    {({ref}) => (
                        <span ref={ref} onClick={this.toggle} className={this.props.textClassName}>
                            {this.props.text}
                        </span>
                    )}
                </Reference>
                {ReactDOM.createPortal(
                    <Popper placement="bottom" positionFixed={true}>
                        {({ref, style, placement, arrowProps}) => (
                            <div ref={ref} style={{...style, ...pointerEvents}} className={cx(
                                "popover",
                                `bs-popover-${placement}`,
                                "fade",
                                {"show": this.state.visible}
                            )}>
                                <div ref={arrowProps.ref} style={arrowProps.style} className="arrow"/>
                                <div className="popover-body">{this.props.children}</div>
                            </div>
                        )}
                    </Popper>,
                    document.querySelector("#modal-root")
                )}
            </Manager>
        );
    }

}
