// ui/control/ReactionSelector.js
import React from 'react';
import PropType from 'prop-types';
import cx from 'classnames';

import { REACTION_EMOJIS } from "api";
import { Twemoji } from "ui/control";
import "./ReactionSelector.css";

const ReactionChoice = ({negative, emoji, invisible, dimmed, onClick}) => {
    const re = !negative ? REACTION_EMOJIS.positive[emoji] : REACTION_EMOJIS.negative[emoji];
    return (
        <div className={cx("choice", {invisible})} onClick={e => onClick(negative, emoji)}>
            <Twemoji key={emoji} code={emoji} title={re ? re.title : ""}/>
            {dimmed && <div className="dimmer" title={re ? re.title : ""}/>}
        </div>
    );
};

export const ReactionSelector = ({negative, reactions, fixedWidth, onClick}) => {
    return (
        <div className="reaction-selector" style={{width: fixedWidth ? "15rem" : "auto"}}>
            {reactions.map(({emoji, invisible, dimmed}) =>
                <ReactionChoice key={emoji} negative={negative} emoji={emoji} invisible={invisible} dimmed={dimmed}
                                onClick={onClick}/>
            )}
        </div>
    );
};

ReactionSelector.propTypes = {
    negative: PropType.bool,
    reactions: PropType.arrayOf(PropType.shape({
        emoji: PropType.number,
        invisible: PropType.bool,
        dimmed: PropType.bool
    })),
    fixedWidth: PropType.bool,
    onClick: PropType.func
};
