// ui/control/RegisteredName.js
import React from 'react';
import PropType from 'prop-types';
import cx from 'classnames';

import { RegisteredName as RegisteredNameParser } from "api";
import "./RegisteredName.css";

const NameGeneration = ({generation, latest}) => (
    latest || <span className="generation">{generation}</span>
);

NameGeneration.propTypes = {
    generation: PropType.number,
    latest: PropType.bool
};

export const RegisteredName = ({ name, latest = false, verified = false, correct = false }) => {
    if (!name) {
        return null;
    }

    const klass = cx(
        "registered-name", {
            "correct": verified && correct,
            "incorrect": verified && !correct
        }
    );
    const parts = RegisteredNameParser.parse(name);
    return (
        <span className={klass}>
            {parts.name}
            <NameGeneration generation={parts.generation} latest={latest} />
        </span>
    );
};

RegisteredName.propTypes = {
    name: PropType.string,
    latest: PropType.bool,
    verified: PropType.bool,
    correct: PropType.bool
};
