// ui/mainmenu/owner/OwnerNavigator.js
import React from 'react';
import { connect } from 'react-redux';
import { Form, withFormik } from 'formik';
import * as yup from 'yup';

import { InputField } from "ui/control/field";
import { Button } from "ui/control";
import { ownerSwitch } from "state/owner/actions";

const OwnerNavigator = ({switching}) => (
    <Form className="form-inline ml-3 mr-3">
        <InputField name="name" horizontal={true} groupClassName="mr-2" autoFocus={true} anyValue/>
        <Button variant="secondary" type="submit" size="sm" loading={switching}>Go</Button>
    </Form>
);

const ownerNavigatorLogic = {

    mapPropsToValues(props) {
        return {
            name: props.ownerName || ""
        };
    },

    validationSchema: yup.object().shape({
        name: yup.string().trim().required("Must not be empty")
    }),

    handleSubmit(values, formik) {
        formik.props.ownerSwitch(values.name.trim());
        formik.setSubmitting(false);
    }

};

export default connect(
    state => ({
        ownerName: state.owner.name,
        switching: state.owner.switching
    }),
    { ownerSwitch }
)(withFormik(ownerNavigatorLogic)(OwnerNavigator));
