// ui/mainmenu/owner/OwnerNavigator.js
import React from 'react';
import { connect } from 'react-redux';
import { Form, withFormik } from 'formik';
import * as yup from 'yup';

import { InputField } from "ui/control/field";
import { Button } from "ui/control";
import { ownerSwitch, ownerSwitchClose } from "state/owner/actions";
import { getNamingNameDetails } from "state/naming/selectors";
import { RegisteredName } from "api/registered-name";

const OwnerNavigator = ({switching, ownerSwitchClose}) => (
    <Form className="form-inline ml-3 mr-3">
        <InputField name="name" horizontal={true} groupClassName="mr-2" autoFocus={true} anyValue
                    onEscape={ownerSwitchClose}/>
        <Button variant="secondary" type="submit" size="sm" loading={switching}>Go</Button>
    </Form>
);

function shortName(name, latest) {
    if (!name) {
        return null;
    }
    return latest ? RegisteredName.parse(name).name : name
}

const ownerNavigatorLogic = {

    mapPropsToValues(props) {
        return {
            name: shortName(props.ownerName, props.latest) || ""
        };
    },

    validationSchema: yup.object().shape({
        name: yup.string().trim().required("Must not be empty")
    }),

    handleSubmit(values, formik) {
        formik.props.ownerSwitch(values.name.trim());
        formik.setSubmitting(false);
    }

};

export default connect(
    state => ({
        ownerName: state.owner.name,
        latest: getNamingNameDetails(state, state.owner.name).latest,
        switching: state.owner.switching
    }),
    { ownerSwitch, ownerSwitchClose }
)(withFormik(ownerNavigatorLogic)(OwnerNavigator));
