// ui/mainmenu/owner/OwnerName.js
import React from 'react';
import PropType from 'prop-types';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import MainMenuTimelineAccessor from "ui/mainmenu/MainMenuTimelineAccessor";
import { isNodeAdmin } from "state/node/selectors";
import { isOwnerNameExpiring } from "state/owner/selectors";
import { RegisteredName } from "ui/control";
import "./OwnerName.css";

class OwnerName extends React.PureComponent {

    static propTypes = {
        name: PropType.string,
        latest: PropType.bool,
        verified: PropType.bool,
        correct: PropType.bool,
        expiring: PropType.bool
    };

    render() {
        if (!this.props.name) {
            return null;
        }
        return (
            <MainMenuTimelineAccessor>
                {(active, href, onClick) => (
                    <span id="owner" className="navbar-text">
                        <a id="owner-name" href={href} onClick={onClick}><RegisteredName {...this.props} /></a>
                        {this.props.expiring &&
                            <FontAwesomeIcon icon="exclamation-triangle" className="owner-name-expiring" />
                        }
                    </span>
                )}
            </MainMenuTimelineAccessor>
        );
    }

}

export default connect(
    state => ({
        ...state.owner,
        expiring: isNodeAdmin(state) && isOwnerNameExpiring(state)
    })
)(OwnerName);
