// ui/mainmenu/owner/OwnerName.js
import React from 'react';
import PropType from 'prop-types';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { isNodeAdmin } from "state/node/selectors";
import { isOwnerNameExpiring } from "state/owner/selectors";
import { RegisteredName } from "ui/control";
import "./OwnerName.css";

class OwnerName extends React.PureComponent {

    static propTypes = {
        name: PropType.string,
        generation: PropType.number,
        latest: PropType.bool,
        verified: PropType.bool,
        correct: PropType.bool,
        expiring: PropType.bool
    };

    render() {
        if (!this.props.name) {
            return null;
        }
        return (
            <span id="owner" className="navbar-text">
                <span id="owner-name"><RegisteredName {...this.props} /></span>
                {this.props.expiring && <FontAwesomeIcon icon="exclamation-triangle" className="owner-name-expiring" />}
            </span>
        );
    }

}

export default connect(
    state => ({
        ...state.owner,
        expiring: isNodeAdmin(state) && isOwnerNameExpiring(state)
    })
)(OwnerName);
